### PR TITLE
feat: Add Papercut module

### PR DESCRIPTION
**Pull request by TechLiam**
## What does this PR do?
Add Testcontainers.Papercut

## Why is it important?
Papercut is a testing SMTP server that also has a nice, easy REST API to get the messages in the inbox.
When doing integration testing it is helpful to be able to send emails and have the ability to check their contents by having this container developers will be able to quickly spin up an SMTP server and check the messages delivered to it without having to send emails to a proper SMTP server and email address.

## How to test this PR
Unit/Integration tests have been added to the solution that checks the container starts and can have emails sent to it. It also checks the helper methods for getting the emails delivered back out of Papercut from the REST API.

### CODE REVIEW

File: src/Testcontainers.Papercut/PapercutContainer.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        client.BaseAddress = new Uri(WebUrl);
+        return client.GetAsync($"/api/messages/{id}/sections/{index}");
+    }
+
+    public Task<HttpResponseMessage> DownloadSectionContent(string id, string contentId)
+    {
+        var client = new HttpClient();
+        client.BaseAddress = new Uri(WebUrl);
+        return client.GetAsync($"/api/messages/{id}/contents/{contentId}");
+    }
```
Comment (HofmeisterAn) [inline src/Testcontainers.Papercut/PapercutContainer.cs]: Is there no Papercut client or, at the very least, a dependency that holds the domain models? Is there a publicly available API document we can refer to? The Testcontainers modules do not implement client functionalities; instead, we focus on spinning up dependencies for various use cases. Can we remove these client functionalities? We can move and keep the necessary domain models to the test and simply send the required HTTP requests from there.

File: src/Testcontainers.Papercut/PapercutBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        Validate();
+        return new PapercutContainer(DockerResourceConfiguration, TestcontainersSettings.Logger);
+    }
+
+    /// <inheritdoc />
+    protected override PapercutBuilder Init()
+    {
+        return base.Init()
+            .WithImage("jijiechen/papercut:latest")
```
Comment (HofmeisterAn) [inline src/Testcontainers.Papercut/PapercutBuilder.cs]: Please use a const and pin the image version, as is done with all the other modules.

File: src/Testcontainers.Papercut/PapercutBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            .WithPortBinding(25,true)
+            .WithPortBinding(37408,true)
```
Comment (HofmeisterAn) [inline src/Testcontainers.Papercut/PapercutBuilder.cs]: Please use a const and a meaningful name that describes the port.

File: src/Testcontainers.Papercut/PapercutBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
+using System.Net;
```
Comment (HofmeisterAn) [inline src/Testcontainers.Papercut/PapercutBuilder.cs]: Move all usings to the global use file.

File: tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return _papercutContainer.DisposeAsync().AsTask();
+    }
+
+    [Fact]
+    [Trait(nameof(DockerCli.DockerPlatform), nameof(DockerCli.DockerPlatform.Linux))]
+    public async Task SendingAnEmail()
+    {
+        //Given
+        var client = new SmtpClient(_papercutContainer.Hostname,_papercutContainer.GetMappedPublicPort(25));
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs]: WDYT about a delegate to access the SmtpPort?

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thanks for your PR. Definitely a cool module. Can we please remove the Papercut client functionalities? This isn't the responsibility of Testcontainers.

File: src/Testcontainers.Papercut/PapercutContainer.cs (hunk 1 above)
Comment (TechLiam) [reply src/Testcontainers.Papercut/PapercutContainer.cs]: Cool, yeah, that makes sense responsibility wise all the container package is doing is creating the container, and interacting with it is outside the scope of this package. I have removed the HTTP Client calls from the package code (moved some/most of it to the tests)

File: src/Testcontainers.Papercut/PapercutBuilder.cs (hunk 2 above)
Comment (TechLiam) [reply src/Testcontainers.Papercut/PapercutBuilder.cs]: Moved it to a const not sure how to pin the version as it's not tagged in the docker hub had a look at how the other packages are doing this and could see that some of the packages e.g. Testcontainers.CosmosDb are using :latest

I'm reasonably new to docker stuff though so if there is a way to do this please let me know I'm happy to add it in

File: src/Testcontainers.Papercut/PapercutBuilder.cs (hunk 3 above)
Comment (TechLiam) [reply src/Testcontainers.Papercut/PapercutBuilder.cs]: Done

File: src/Testcontainers.Papercut/PapercutBuilder.cs (hunk 4 above)
Comment (TechLiam) [reply src/Testcontainers.Papercut/PapercutBuilder.cs]: Done

File: tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs (hunk 5 above)
Comment (TechLiam) [reply tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs]: I agree this is better so have added it in

File: src/Testcontainers.Papercut/PapercutBuilder.cs (hunk 2 above)
Comment (HofmeisterAn) [reply src/Testcontainers.Papercut/PapercutBuilder.cs]: > not sure how to pin the version as it's not tagged

👍 Okay, then `latest` is fine. I didn't check the image tags. Thanks for checking.

**Comment by TechLiam** [comment]
@HofmeisterAn is there anything else or is this PR good to merge now?

**Comment by HofmeisterAn** [comment]
> @HofmeisterAn is there anything else or is this PR good to merge now?

I've made some improvements. WDYT? If you're good with it, we are ready to merge.

File: tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs [hunk 6]
```diff
@@ -16,33 +16,57 @@ public Task DisposeAsync()
...
+            messages = jsonDocument.RootElement.GetProperty("messages").Deserialize<Message[]>();
+        }
+        while (messages.Length == 0);
+
+        // Then
+        Assert.NotEmpty(messages);
+        Assert.Equal(subject, messages[0].Subject);
+    }
+
+    private readonly struct Message
```
Comment (TechLiam) [inline tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs:49]: I like this struct hadn't thought about doing it with that good call

**Review by HofmeisterAn — DISMISSED** [review]
I guess you're good with the recent changes. Thanks for the pull request. Great first Testcontainers contribution 👍. I will merge it later the day.

**Pull request merged**